Build issue with MPI on Windows

The following trivial example errors on Windows using VS Build Tools 2017

```C++
#include <amgcl/amg.hpp>
#include <amgcl/make_solver.hpp>
#include <amgcl/backend/builtin.hpp>
#include <amgcl/adapter/ublas.hpp>
#include <amgcl/coarsening/smoothed_aggregation.hpp>
#include <amgcl/relaxation/chebyshev.hpp>
#include <amgcl/solver/bicgstab.hpp>
#include <amgcl/value_type/static_matrix.hpp>
#include <amgcl/adapter/block_matrix.hpp>

#include <amgcl/mpi/direct_solver/runtime.hpp>
#include <amgcl/mpi/coarsening/runtime.hpp>
#include <amgcl/mpi/relaxation/runtime.hpp>
#include <amgcl/mpi/partition/runtime.hpp>
#include <amgcl/mpi/amg.hpp>
#include <amgcl/mpi/make_solver.hpp>
#include <amgcl/solver/runtime.hpp>
#include <amgcl/profiler.hpp>

int main()
{
}
```

with the following cmake file:

```cmake
cmake_minimum_required (VERSION 3.11)
project (Tutorial)
add_executable(Tutorial tutorial.cxx)

add_definitions( -DBOOST_ALL_NO_LIB )

find_package(MPI REQUIRED)
find_package(Boost COMPONENTS serialization REQUIRED)

add_subdirectory("amgcl")

target_include_directories(
    Tutorial
        SYSTEM PRIVATE amgcl
		SYSTEM PRIVATE ${Boost_INCLUDE_DIRS}
        SYSTEM PRIVATE ${MPI_CXX_INCLUDE_DIRS})

message(${Boost_LIBRARY_DIRS})

target_link_libraries(
    Tutorial
        PRIVATE ${CMAKE_THREAD_LIBS_INIT}
        PRIVATE ${MPI_CXX_LIBRARIES}
        PRIVATE ${Boost_LIBRARIES}
        PRIVATE amgcl::amgcl)

set_property(TARGET Tutorial PROPERTY CXX_STANDARD 17)
```

And the following two error messages:

```
c:\users\XXX\PROJECT\amgcl\amgcl\mpi\util.hpp(104): error C2766: explicit specialization; 'amgcl::mpi::datatype_impl<__int64,void>' has already been defined [C:\Users\XXX\PROJECT\build\Tutorial.vcxproj]
c:\users\XXX\PROJECT\amgcl\amgcl\mpi\util.hpp(111): error C2766: explicit specialization; 'amgcl::mpi::datatype_impl<unsigned __int64,void>' has already been defined
```

I am using the current master branch and the relevant code is:

https://github.com/ddemidov/amgcl/blob/3febe6853e4a3f07aabb31619a44ad3258d90e69/amgcl/mpi/util.hpp#L47-L111

With this branch I get the following output:

```
[build]          C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.15.26726\bin\HostX86\x64\CL.exe /c /IC:\Users\XXX\YYY\amgcl /IC:\Users\XXX\YYY\SYSTEM /I"C:\Program Files (x86)\Microsoft SDKs\MPI\Include" /IC:\local\boost_1_67_0 /Zi /nologo /W3 /WX- /diagnostics:classic /Od /Ob0 /D WIN32 /D _WINDOWS /D NOMINMAX /D _USE_MATH_DEFINES /D _VARIADIC_MAX=10 /D _SCL_SECURE_NO_WARNINGS /D AMGCL_ASYNC_SETUP /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:fast /Zc:wchar_t /Zc:forScope /Zc:inline /GR /openmp /Fo"Tutorial.dir\Debug\\" /Fd"Tutorial.dir\Debug\vc141.pdb" /Gd /TP /wd4715 /FC /errorReport:queue  /bigobj C:\Users\XXX\YYY\tutorial.cxx
[build]          tutorial.cxx
[build]      3>c:\users\XXX\YYY\amgcl\amgcl\util.hpp(314): warning C4244: 'initializing': conversion from 'size_t' to 'double', possible loss of data [C:\Users\XXX\YYY\build\Tutorial.vcxproj]
[build]      3>c:\users\XXX\YYY\amgcl\amgcl\mpi\util.hpp(101): error C2039: 'type': is not a member of 'std::enable_if<false,void>' [C:\Users\XXX\YYY\build\Tutorial.vcxproj]
[build]          c:\program files (x86)\microsoft visual studio\2017\buildtools\vc\tools\msvc\14.15.26726\include\type_traits(227): note: see declaration of 'std::enable_if<false,void>'
[build]      3>c:\users\XXX\YYY\amgcl\amgcl\mpi\util.hpp(102): error C2146: syntax error: missing '>' before identifier 'type' [C:\Users\XXX\YYY\build\Tutorial.vcxproj]
[build]      3>c:\users\XXX\YYY\amgcl\amgcl\mpi\util.hpp(109): error C2039: 'type': is not a member of 'std::enable_if<false,void>' [C:\Users\XXX\YYY\build\Tutorial.vcxproj]
[build]          c:\program files (x86)\microsoft visual studio\2017\buildtools\vc\tools\msvc\14.15.26726\include\type_traits(227): note: see declaration of 'std::enable_if<false,void>'
[build]      3>c:\users\XXX\YYY\amgcl\amgcl\mpi\util.hpp(110): error C2146: syntax error: missing '>' before identifier 'type' [C:\Users\XXX\YYY\build\Tutorial.vcxproj]
[build]      3>Done Building Project "C:\Users\XXX\YYY\build\Tutorial.vcxproj" (default targets) -- FAILED.
[build]      1>Done Building Project "C:\Users\XXX\YYY\build\ALL_BUILD.vcxproj" (default targets) -- FAILED.
[build] 
[build] Build FAILED.

Ok. I updated the example. It now compiles. There is one warning though. I don't know if you want to do anything about it:

    c:\XXX\YYY\test_copysign\amgcl\amgcl\util.hpp(314): warning C4244: 'initializing': conversion from 'size_t' to 'double', possible loss of data

--- a/amgcl/mpi/util.hpp
+++ b/amgcl/mpi/util.hpp
@@ -97,14 +97,17 @@ struct datatype_impl<unsigned long long> {
 };
 
 template <>
-struct datatype_impl<ptrdiff_t>
-    : std::conditional<
+struct datatype_impl<ptrdiff_t,
+    typename std::enable_if<!std::is_same<ptrdiff_t, long long>::value && !std::is_same<ptrdiff_t, int>::value>::type
+    > : std::conditional<
         sizeof(ptrdiff_t) == sizeof(int), datatype_impl<int>, datatype_impl<long long>
         >::type
 {};
 
 template <>
-struct datatype_impl<size_t>
+struct datatype_impl<size_t,
+    typename std::enable_if<!std::is_same<size_t, unsigned long long>::value && !std::is_same<ptrdiff_t, unsigned int>::value>::type
+    >
     : std::conditional<
         sizeof(size_t) == sizeof(unsigned), datatype_impl<unsigned>, datatype_impl<unsigned long long>
         >::type

--- a/amgcl/mpi/util.hpp
+++ b/amgcl/mpi/util.hpp
@@ -96,17 +96,25 @@ struct datatype_impl<unsigned long long> {
     static MPI_Datatype get() { return MPI_UNSIGNED_LONG_LONG; }
 };
 
-template <>
-struct datatype_impl<ptrdiff_t,
-    typename std::enable_if<!std::is_same<ptrdiff_t, long long>::value && !std::is_same<ptrdiff_t, int>::value>::type
+template <typename T>
+struct datatype_impl<T,
+    typename std::enable_if<
+        std::is_same<T, ptrdiff_t>::value &&
+        !std::is_same<ptrdiff_t, long long>::value &&
+        !std::is_same<ptrdiff_t, int>::value
+        >::type
     > : std::conditional<
         sizeof(ptrdiff_t) == sizeof(int), datatype_impl<int>, datatype_impl<long long>
         >::type
 {};
 
-template <>
-struct datatype_impl<size_t,
-    typename std::enable_if<!std::is_same<size_t, unsigned long long>::value && !std::is_same<ptrdiff_t, unsigned int>::value>::type
+template <typename T>
+struct datatype_impl<T,
+    typename std::enable_if<
+        std::is_same<T, size_t>::value &&
+        !std::is_same<size_t, unsigned long long>::value &&
+        !std::is_same<ptrdiff_t, unsigned int>::value
+        >::type
     >
     : std::conditional<
         sizeof(size_t) == sizeof(unsigned), datatype_impl<unsigned>, datatype_impl<unsigned long long>

